RFC-0022-Model-Partitioning-in-Pipeline-Parallelism.md

```diff
@@ -0,0 +1,255 @@
+
+
+## Relevant works
+
+How is the industry dealing with model partitioning for pipeline parallelism so far? There’re plenty exploration works going on from different pipeline parallelism implementations, and some of them produces pretty nice performance improvements. 
+
+***FairScale/torchgpipe:***
+FairScale and torchgpipe use similar/same approaches when partitioning the model, their partition APIs:
+
+*  `Pipe(..., balance: List[int],...)` use `balance` to partition the model during construction
+* two utils: `balance_by_time` and `balance_by_sizes` to generate the `balance` list automatically based on execution time or memory sizes (parameter + optim states)
+
+Partition Algorithm: use “Block Partition of Sequences” algorithm to find a balanced partition base on the costs. 
+Note they only allow to partition on the top-level modules of the Sequential list based on the costs (memory/time)
+
+***SageMaker***
+Major logic is under `model_partition.py` and the basic ideas:
+
+* `ModuleParitioner` to drive the model partition progress base on trace results 
+* `ModulePartitioner.partition` create a tree of `ModuleNodes`, populate the costs (execution time + memory cost) and run the partition algorithm, return a `Dict[Module, device_id]`
+
+Partition Algorithm: SageMaker provided an automated model parallelism by their own proposed model partition algorithm based on BFS + DP-based method for device allocations, plus reallocation using d'Hondt method
+
+***DeepSpeed:***
+Provide several mechanisms for partition the model across GPUs with `partition_method`:
+
+* `partition_method="parameters"` (default): balances the number of trainable parameters on each pipeline stage.
+* `partition_method="type:[regex]"`: balances layers whose class names match [regex].
+* `partition_method="uniform"`: balances the number of layers per stage.
+
+Partition Algorithm: For partition_method = “parameters”, DeepSpeed use the mechanism of counting the layer’s parameters size, and do a binary search (find the smallest weight of the heaviest partition) to find a balanced partition.
+
+***PipeDream***
+More on asynchronous training pipeline planning, like parameter server approaches, the algorithm itself also uses Dynamic Programming, but the states is highly correlated with the communication and synchronization costs, which is not a good choice for us (unless we consider async training in the future). 
+
+***DAPPLE***
+
+DAPPLE combines DDP with pipeline parallelism to form a bigger search space and use a more efficient schedule if possible. Their partition approach is a DP-based algorithm, it first tries to find the “pivotal” stage, then optimize the overall latency,  the “overall latency” optimization here tries to reduce the bubbles in the pipeline as small as possible.
+
+
```

Comment:
Here too Varuna that I linked too earlier provides a partitioning mechanism using cutPoints. see their paper : https://arxiv.org/abs/2111.04007 for details.